doc/src/manual/constructors.md

```diff
@@ -145,6 +149,7 @@ julia> struct T1
 julia> struct T2
            x::Int64
            T2(x) = new(x)
```

Comment:
If we want to mimic the actual constructors that are created, then your addition plus

```suggestion
           T2(x) = T2(convert(Int64, x)
```

would do the trick.

Reply:
See https://github.com/JuliaLang/julia/pull/51253#discussion_r1323222154 above.  I think what I wrote is correct.

Reply:
That's because you manually called `new` from `S2(x::Int)`. The default calls `convert`. Let's look at the authoritative source:

```julia
julia> Meta.lower(Main, quote
       struct S1
           x::Float64
       end
       end)
:($(Expr(:thunk, CodeInfo(
     @ REPL[1]:2 within `top-level scope`
1 ──       global S1
│          const S1
│    %3  = Core.svec()
│    %4  = Core.svec(:x)
│    %5  = Core.svec()
│          #1#S1 = Core._structtype(Main, :S1, %3, %4, %5, false, 1)
│          Core._setsuper!(#1#S1, Core.Any)
│    %8  = $(Expr(:isdefined, :S1))
└───       goto #6 if not %8
2 ── %10 = S1
│    %11 = Core._equiv_typedef(%10, #1#S1)
└───       goto #4 if not %11
3 ──       #1#S1 = %10
└───       goto #5
4 ── %15 = #1#S1
└───       S1 = %15
5 ┄─       goto #7
6 ── %18 = #1#S1
└───       S1 = %18
7 ┄─ %20 = #1#S1
│    %21 = Core.svec(Float64)
│          Core._typebody!(%20, %21)
│          global S1
│    %24 = Core.Any === Float64
└───       goto #9 if not %24
8 ──       goto #10
9 ── %27 = Core.apply_type(Core.Type, S1)
│    %28 = Core.svec(%27, Float64)
│    %29 = Core.svec()
│    %30 = Core.svec(%28, %29, $(QuoteNode(:(#= REPL[1]:3 =#))))
└───       $(Expr(:method, nothing, :(%30), CodeInfo(
    @ REPL[1]:3 within `none`
1 ─ %1 = %new(S1, x)
└──      return %1
)))
10 ┄ %32 = Core.apply_type(Core.Type, S1)
│    %33 = Core.svec(%32, Core.Any)
│    %34 = Core.svec()
│    %35 = Core.svec(%33, %34, $(QuoteNode(:(#= REPL[1]:3 =#))))
│          $(Expr(:method, nothing, :(%35), CodeInfo(
    @ REPL[1]:3 within `none`
1 ─ %1 = S1
│   %2 = Core.fieldtype(%1, 1)
│   %3 = Base.convert(%2, x)
│   %4 = %new(%1, %3)
└──      return %4
)))
└───       return nothing
))))
```

This is a little hard to read, but:
- up through `%26` (not displayed, but it's the `goto #10`) is the type-definition
- `%27-%31` is the method definition that calls `new`
- `%32` onward is the method definition that calls `convert`

You're being fooled by the fact that `@code_typed` allows inlining. Try `@code_typed optimize=false S1(1)`, which only performs type-inference and does not inline the `convert` call:

```julia
julia> @code_typed optimize=false S1(1)
CodeInfo(
1 ─ %1 = Main.S1::Core.Const(S1)
│   %2 = Core.fieldtype(%1, 1)::Core.Const(Float64)
│   %3 = Base.convert(%2, x)::Float64
│   %4 = %new(%1, %3)::S1
└──      return %4
) => S1
```

Reply:
Now that I see the source of the confusion, I took the liberty of editing your title. The existing docs are not wrong, but they could use amplification.

Reply:
@timholy I tried both ways.  Specifically, I tried both
```julia
struct S1a
    x::Float64
end
```
and
```julia
struct S1a
    x::Float64
    S1a(x) = new(x)
    S1a(x::Float64) = new(convert(Float64, x))
end
```
and both produce precisely the same result for `@code_lowered S1a(1)`.  So it seems the conversion is not required in the translated "as if" source.

Thank you for correcting my title.  Can you please direct me to where the existing specification shows the presence of *two* inner constructors when one writes the following?  If it isn't anywhere in the spec, then it is a bug and not a correction.  If it is there but not clear, then yes I would call it a correction.

```julia
struct S1a
    x::Float64
end
```

Reply:
```julia
julia> struct S1a
           x::Float64
       end

julia> println(methods(S1a))
# 2 methods for type constructor:
[1] S1a(x::Float64) in Main at REPL[1]:2
[2] S1a(x) in Main at REPL[1]:2
```

Reply:
I'm not disagreeing that two methods are created. I'm objecting to a laser-like focus on the distinction between inner and outer.

Reply:
@timholy I've verified that the translation I documented has identical behavior to the default constructor for non-parametric types, including the behavior with respect to conversions and whether or not it is an inner constructor.  For parametric types the translation is documented in a later section of this document, and I have not changed that section.

If your point is that there is a conversion implied by a call to `new` (even the call produced by the translation in my specification), and that conversion is not currently documented, I agree. Would documenting that help?